scheds/rust/scx_rustlite/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,370 @@
+	 */
+	if (is_usersched_task(p))
+		return;
+
+        /*
+	 * Dispatch per-cpu kthreads on the local FIFO directly from the
+	 * kernel.
+         */
+	if (is_kthread(p) && p->nr_cpus_allowed == 1) {
+		bpf_printk("dispatch (local): pid=%d", p->pid);
+		dispatch_task_local(p, enq_flags | SCX_ENQ_LOCAL);
+		return;
+	}
+
+	/*
+	 * Other tasks can be added to the @enqueued list and they will be
+	 * processed by the user-space scheduler.
+	 *
+	 * If the @enqueued list is full (user-space scheduler is congested)
+	 * tasks will be dispatched directly from the kernel to the global
+	 * FIFO.
+	 */
+	task.pid = p->pid;
+	task.vtime = task_vtime(p);
+	bpf_printk("enqueue: pid=%d vtime=%llu", task.pid, task.vtime);
+	if (bpf_map_push_elem(&enqueued, &task, 0)) {
+		bpf_printk("dispatch (global): pid=%d", task.pid);
+		dispatch_task_global(p, enq_flags);
+		return;
+	}
+	__sync_fetch_and_add(&nr_enqueues, 1);
+
+	/*
+	 * Task was sent to user-space correctly, now we can wake-up the
+	 * user-space scheduler.
+	 */
+	usersched_needed = true;
+}
+
+/* Dispatch the user-space scheduler */
+static void dispatch_user_scheduler(void)
+{
+	struct task_struct *p;
+
+	if (!usersched_needed)
+		return;
+	usersched_needed = false;
+
+	p = bpf_task_from_pid(usersched_pid);
+	if (!p) {
+		scx_bpf_error("Failed to find usersched task %d", usersched_pid);
+		return;
+	}
+	bpf_printk("dispatch (scheduler): pid=%d", usersched_pid);
+	scx_bpf_dispatch(p, SCX_DSQ_GLOBAL, slice_ns, 0);
+	bpf_task_release(p);
+}
+
+/* Dispatch tasks that are ready to run */
+void BPF_STRUCT_OPS(rustlite_dispatch, s32 cpu, struct task_struct *prev)
+{
+	/* Check if the user-space scheduler needs to run */
+	dispatch_user_scheduler();
+
+	/*
+	 * Consume all tasks from the dispatched list and immediately dispatch
+	 * them to the global FIFO (the proper ordering has been already
+	 * determined by the user-space scheduler).
+	 *
+	 * A trace of the dispatched tasks can be seen in
+	 * /sys/kernel/debug/tracing/trace_pipe.
+	 */
+	bpf_repeat(MAX_ENQUEUED_TASKS) {
+		struct task_struct *p;
+		struct task_ctx task;
+
+		if (!scx_bpf_dispatch_nr_slots())
+			break;
+		if (bpf_map_pop_elem(&dispatched, &task))
+			break;
+		p = bpf_task_from_pid(task.pid);
+		if (!p)
+			continue;
+		bpf_printk("dispatch: pid=%d vtime=%llu", task.pid, task.vtime);
+		dispatch_task_global(p, 0);
+		bpf_task_release(p);
+	}
+}
+
+/* Select the target CPU for a task that is ready to run */
+s32 BPF_STRUCT_OPS(rustlite_select_cpu, struct task_struct *p, s32 prev_cpu,
+		   u64 wake_flags)
+{
+	s32 cpu;
+
+	/*
+	 * Keep using the same CPU if the task can only run there or if the CPU
+	 * is idle.
+	 */
+	if (p->nr_cpus_allowed == 1 ||
+	    scx_bpf_test_and_clear_cpu_idle(prev_cpu))
+		return prev_cpu;
+
+	/*
+	 * Otherwise migrate to the next idle CPU, if available (this strategy
+	 * is not ideal, since it assumes a migration cost of zero).
+	 */
+	cpu = scx_bpf_pick_idle_cpu(p->cpus_ptr, 0);
```

Comment:
I'll probably mimic what `scx_userland` is doing, as suggested by @Decave. 
In perspective I would also like to offload the logic of `ops.select_cpu()` to userspace as well.

Reply:
This probably needs better documentation but `.select_cpu()` is purely an optimization. `.enqueue()` and `.dispatch()` have complete control over which cpu a given task runs on. The only thing `.select_cpu()` brings to the table is that if its verdict and the CPU `.enqueue()` or `.dispatch()` ends up choosing for the task matches the one returned from `.select_cpu()`, we can avoid a lock dancing to hot-migrate the task for execution thus reducing the overhead by a small amount.

Here, because all non-trivial scheduling decisions are punted to userspace, `.select_cpu()` doesn't have any way of predicting what the eventual CPU selection is going to be and thus whatever work it does is wasted. Also, `.select_cpu()` can't be offloaded to userland because it's in synchronous wakeup path. Besides, it doesn't make whole lot of sense anyway. The overhead it's saving when it successfully predicts the CPU the task is going to run on is fairly small. It's likely impossible to offset the travel to userspace even if we can offload it to userland.

Reply:
Hm... so maybe we can just rely on `select_cpu_dfl()` and get rid of `ops.select_cpu()` in this scheduler.

Or what about offloading cpu selection to user-space without using `ops.select_cpu()`?

Basically add some simple logic to the Rust code that would update `.target_cpu` in the task context and then always use `scx_bpf_dispatch(p, SCX_DSQ_LOCAL_ON | target_cpu, slice_ns, 0)` in `ops.dispatch()`. Would that work?

Reply:
I think the right thing to do is having a really dumb implementation which just returns @prev_cpu. That's cheaper than the default implementation and likely the best behavior anyway.

Reply:
BTW, adding `.taget_cpu` can work too; however, I'm skeptical it'd be worth the complication. It's not a huge optimization. If we're bouncing through userspace to make a scheduler decision anyway, we aren't going to notice this micro optimization.

Reply:
ok makes sense, let's keep it simple and just return @prev_cpu. Thanks for the review and all the good suggestions @htejun !